test/jdk/com/sun/jdi/SuspendAfterDeath.java

```diff
@@ -0,0 +1,135 @@
+        } else {
+            thread = Thread.ofPlatform().name(THREAD_NAME).start(() -> { });
+        }
+        thread.join();
+        done();
+    }
+}
+
+public class SuspendAfterDeath extends TestScaffold {
+    private volatile ThreadReference thread;
+    private volatile boolean breakpointReached = false;
+
+    SuspendAfterDeath(String args[]) {
+        super(args);
+    }
+
+    public static void main(String[] args) throws Exception {
+        new SuspendAfterDeath(args).startTests();
+    }
+
+    @Override
+    public void threadDied(ThreadDeathEvent event) {
+        ThreadReference eventThread = event.thread();
+        if (eventThread.name().equals(SuspendAfterDeathTarg.THREAD_NAME)) {
+            System.out.println("Target thread died, thread=" + eventThread);
+            thread = eventThread;
+        }
+    }
+
+    @Override
+    public void breakpointReached(BreakpointEvent event) {
+        ThreadReference eventThread = event.thread();
+        System.out.println("Breakpoint, thread=" + eventThread);
+        if (thread == null) {
+            failure("FAILED: got Breakpoint event before ThreadDeath event.");
+        }
+        breakpointReached = true;
+        /* Suspend the thread. This is being done after the thread has exited. */
+        thread.suspend();
+    }
+
+    @Override
+    public void connect(String args[]) {
+        String mainWrapper = System.getProperty("main.wrapper");
+        if ("Virtual".equals(mainWrapper)) {
+            List argList = new ArrayList(Arrays.asList(args));
+            argList.add("Virtual");
+            args = (String[]) argList.toArray(args);
+        }
+        super.connect(args);
+    }
+
+    @Override
+    protected void runTests() throws Exception {
+        BreakpointEvent bpe = startToMain("SuspendAfterDeathTarg");
+        EventRequestManager erm = vm().eventRequestManager();
+
+        // listener for ThreadDeathEvent captures reference to the thread
+        ThreadDeathRequest request1 = erm.createThreadDeathRequest();
+        request1.enable();
+
+        // listener for BreakpointEvent attempts to suspend the thread
+        ReferenceType targetClass = bpe.location().declaringType();
+        Location loc = findMethod(targetClass, "done", "()V").location();
+        BreakpointRequest request2 = erm.createBreakpointRequest(loc);
+        request2.setSuspendPolicy(EventRequest.SUSPEND_EVENT_THREAD);
+        request2.enable();
+
+        listenUntilVMDisconnect();
+
+        if (thread == null) {
+            failure("FAILED: never got ThreadDeath event for target thread.");
+        }
+        if (!breakpointReached) {
+            failure("FAILED: never got Breakpoint event for target thread.");
+        }
+
+        if (!testFailed) {
+            println("SuspendAfterDeath: passed");
+        } else {
+            throw new Exception("SuspendAfterDeath: failed");
+        }
+    }
+}
```

Comment:
I wonder if we need to add `@run main/othervm -Dmain.wrapper=Virtual SuspendAfterDeath` to the test description so that running the jdk_jdi test group will exercise the issue. As it stands, I think it would require. test run with -Dmain.wrapper=Virtual to exercise this code.

Reply:
Yes, it requires `-Dmain.wrapper=Virtual`, and that was intentional. My thinking was that we don't have any com/sun/jdi tests that do any virtual thread testing unless run with `-Dmain.wrapper=Virtual`, so why should this test be any different? If I made it do virtual thread testing without `-Dmain.wrapper=Virtual`, then the `-Dmain.wrapper=Virtual` testing becomes redundant.

Reply:
The test uses a platform thread by default so it's not testing the issue in JDK-8287847. The suggestion is to have it re-run with the system property set, like this:
```
@run main/othervm SuspendAfterDeath 
@run main/othervm -Dmain.wrapper=Virtual SuspendAfterDeath 
```
I accept this may be different to the existing com/sun/jdi tests but they pre-date virtual threads.

Reply:
The @run suggestion doesn't work. There are two issues. First, it does not trigger jtreg to use TestScaffold.main(), which is what normally happens when you use TEST_VM_OPTS="-Dmain.wrapper=Virtual". So the end result is that the test is run with a platform thread.

The other issue is that if the test run is done with TEST_VM_OPTS="-Dmain.wrapper=Virtual", then TestScaffold.main() cannot handle any arguments added to the @run other than the main class name. You end up with CNFE for the first argument that appears after "SuspendAfterDeath".

The above issues aside, I still don't understand why we would want to do any virtual thread testing with com/sun/jdi tests when not using TEST_VM_OPTS="-Dmain.wrapper=Virtual". It's not how we've tested JDI support for virtual thread in the past. We've always relied on explicit virtual thread test runs using TEST_VM_OPTS="-Dmain.wrapper=Virtual". Same is true of the nsk jdi, jdb, and jdwp tests. If you really want this test handled differently because it was written to test a specific bug, then it probably does not belong in com/sun/jdi, which means it can't rely on TestScaffold.

One other thing is that you don't really need -Dmain.wrapper=Virtual to make this test do the testing on a virtual thread. It can just be the default mode like you used to have when you first wrote it. However, I ran into problems with that when also using  TEST_VM_OPTS="-Dmain.wrapper=Virtual" due to conflicts in argument setup between the test and TestScaffold. However, with the changes now in parseArgs() to check test.enable.preview, it might not be necessary for the test itself to add --enable-preview, so maybe that conflict will go away. But I still don't think this is the approach we should take (see my previous paragraph).

Reply:
The testing in OpenJDK does not run these tests with -Dmain.wrapper=Virtual so this is why this test is not testing the scenario that this bug is about.  We can make it work and have it run twice, or we can rename it to something like "SuspendVirtualThreadAfterDeath" and drop the testing with a platform thread. In time I expect there will be more tests in the jdk_jdi test group that exercise virtual threads.




It can be made to work but I would be okay with dropping the testing of platform thread from this test. 


The reason for asking for changes here is the test, as presented, doesn't exercise virtual threads with any of the tiers defined

Reply:
I've updated the test to always run in both virtual thread and platform thread mode. The test itself does not care about the `-Dmain.wrapper=Virtual` setting and ignores it. TestScaffold does care and will run the main test thread as a virtual thread if `-Dmain.wrapper=Virtual` is used, but the test itself will spawn a virtual thread or platform thread independent of the wrapper setting.